servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/RepeatWhenSingleTest.java

```diff
@@ -0,0 +1,272 @@
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? completed() : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(1);
+        assertThat(subscriber.takeOnNext(), equalTo(1));
+        s.request(-1);
+        assertThat(subscriber.awaitOnError(), instanceOf(IllegalArgumentException.class));
+    }
+
+    @Test
+    void singleRepeatAllDemandUpfront() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? completed() : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(2);
+        assertThat(subscriber.takeOnNext(2), contains(1, 2));
+        subscriber.awaitOnComplete();
+    }
+
+    @Test
+    void manyRepeatAllDemandUpfront() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount < 5 ? completed() : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(5);
+        assertThat(subscriber.takeOnNext(5), contains(1, 2, 3, 4, 5));
+        subscriber.awaitOnComplete();
+    }
+
+    @Test
+    void singleRepeatDelayedDemand() {
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? executor.submit(() -> {
+                    try {
+                        latch.await();
+                    } catch (InterruptedException e) {
+                        throw new AssertionError(e);
+                    }
+                }) : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(2);
+        assertThat(subscriber.takeOnNext(), equalTo(1));
+        assertThat(subscriber.pollOnNext(10, MILLISECONDS), nullValue());
+        latch.countDown();
+        assertThat(subscriber.takeOnNext(), equalTo(2));
+        subscriber.awaitOnComplete();
+    }
+
+    @Test
+    void singleDelayedRepeat() {
+        AtomicInteger value = new AtomicInteger();
+        toSource(Single.defer(() -> Single.succeeded(value.incrementAndGet()))
+                .repeatWhen((repeatCount, v) -> repeatCount == 1 ? completed() : failed(new DeliberateException())))
+                .subscribe(subscriber);
+        Subscription s = subscriber.awaitSubscription();
+        s.request(2);
```

Comment:
Consider also testing when `s.request(1)` is made multiple times in a row and from inside `onNext`.